File: python/remove-duplicates-from-sorted-list-ii.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy_head = prev = ListNode(0, head)
        curr = head

        while curr:
            val = curr.val
            if curr.next and curr.next.val == val:
                curr = curr.next
                while curr and curr.val == val:
                    curr = curr.next
            else:
                prev.next = curr
                prev, curr = curr, curr.next
        prev.next = None
        return dummy_head.next

File: go/remove_duplicates_from_sorted_list_ii.go
package leetcode

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(n)
// Space complexity: O(1)
func deleteDuplicates(head *ListNode) *ListNode {
	dummyHead := &ListNode{0, head}
	prev := dummyHead
	curr := head

	for curr != nil {
		val := curr.Val
		if curr.Next != nil && curr.Next.Val == val {
			curr = curr.Next
			for curr != nil && curr.Val == val {
				curr = curr.Next
			}
		} else {
			prev.Next = curr
			prev, curr = curr, curr.Next
		}
	}
	prev.Next = nil
	return dummyHead.Next
}
